pkg/deviceclaimingserver/grpc_end_devices.go

```diff
@@ -67,21 +66,28 @@ var (
 )
 
 // Claim implements EndDeviceClaimingServer.
-func (edcs *endDeviceClaimingServer) Claim(ctx context.Context, req *ttnpb.ClaimEndDeviceRequest) (ids *ttnpb.EndDeviceIdentifiers, err error) {
-	for _, edcs := range edcs.DCS.endDeviceClaimingUpstreams {
+func (srv *endDeviceClaimingServer) Claim(ctx context.Context, req *ttnpb.ClaimEndDeviceRequest) (ids *ttnpb.EndDeviceIdentifiers, err error) {
+	for _, edcs := range srv.DCS.endDeviceClaimingUpstreams {
 		var joinEUI types.EUI64
 		if authenticatedIDs := req.GetAuthenticatedIdentifiers(); authenticatedIDs != nil {
 			joinEUI = req.GetAuthenticatedIdentifiers().JoinEui
 		} else if qrCode := req.GetQrCode(); qrCode != nil {
-			data, err := qrcode.Parse(qrCode)
+			conn, err := srv.DCS.GetPeerConn(ctx, ttnpb.ClusterRole_QR_CODE_GENERATOR, nil)
+			if err != nil {
+				return nil, err
+			}
+			client := ttnpb.NewEndDeviceQRCodeGeneratorClient(conn)
+			data, err := client.Parse(ctx, &ttnpb.ParseEndDeviceQRCodeRequest{
+				QrCode: qrCode,
+			})
```

Comment:
Is there no auth here at all, like not even checking whether there is any authentication?

If the API call is open, does it matter if we use cluster auth here to bypass the rate limiter?

Reply:
The value that's passed as input (QR code contents) need to more secure than the data that comes out of this RPC so I don't think auth is required here. I've removed the cluster auth since that's indeed not necessary.

Reply:
I meant that if we don't use any auth, we can't do rate limiting other than on originating IP.